pyogrio/_io.pyx

```diff
@@ -464,19 +483,25 @@ cdef get_feature_count(OGRLayerH ogr_layer, int force):
     """
 
     cdef OGRFeatureH ogr_feature = NULL
-    cdef int feature_count = OGR_L_GetFeatureCount(ogr_layer, force)
+    cdef int feature_count
+
+    with nogil:
+        feature_count = OGR_L_GetFeatureCount(ogr_layer, force)
 
     # if GDAL refuses to give us the feature count, we have to loop over all
     # features ourselves and get the count.  This can happen for some drivers
     # (e.g., OSM) or if a where clause is invalid but not rejected as error
     if force and feature_count == -1:
         # make sure layer is read from beginning
-        OGR_L_ResetReading(ogr_layer)
+        with nogil:
+            OGR_L_ResetReading(ogr_layer)
 
         feature_count = 0
         while True:
             try:
-                ogr_feature = check_pointer(OGR_L_GetNextFeature(ogr_layer))
+                with nogil:
+                    ogr_feature = OGR_L_GetNextFeature(ogr_layer)
```

Comment:
Since we are in a loop here for counting the features, this means for a file with 10,000 rows, we are going to release/acquire the GIL 10,000 times? 
(if I am not misreading the code, that might not be optimal)